frontend/packages/schema-editor/src/SchemaEditorApp.tsx

```diff
@@ -1,28 +1,83 @@
-import React, { useCallback, useMemo } from 'react';
+import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
 import { Provider } from 'react-redux';
 import './App.css';
-import { SchemaEditor } from './components/SchemaEditor';
 
 import { store } from './store';
 import '@digdir/design-system-tokens/brand/altinn/tokens.css';
 import { SchemaEditorAppContext } from '@altinn/schema-editor/contexts/SchemaEditorAppContext';
 import { JsonSchema } from 'app-shared/types/JsonSchema';
 import { buildJsonSchema, buildUiSchema, UiSchemaNodes } from '@altinn/schema-model';
+import { AUTOSAVE_DEBOUNCE_INTERVAL } from 'app-shared/constants';
+import { DatamodelMetadata } from 'app-shared/types/DatamodelMetadata';
 
 export type SchemaEditorAppProps = {
-  modelName?: string;
+  children?: React.ReactNode;
+  datamodels: DatamodelMetadata[];
+  modelPath?: string;
   jsonSchema: JsonSchema;
-  save: (schema: JsonSchema) => void;
-}
+  save: (args: { modelPath: string; model: JsonSchema }) => void;
+};
+
+export function SchemaEditorApp({
+  children,
+  datamodels,
+  modelPath,
+  jsonSchema,
+  save,
+}: SchemaEditorAppProps) {
+  const autoSaveTimeoutRef = useRef(undefined);
+  const [model, setModel] = useState(() => (jsonSchema ? buildUiSchema(jsonSchema) : []));
+  const prevModelPathRef = useRef(modelPath);
+  const prevModelRef = useRef(model);
+
+  const saveInternalModel = useCallback(
+    (newModel: UiSchemaNodes, saveAfterMs: number = AUTOSAVE_DEBOUNCE_INTERVAL) => {
+      prevModelRef.current = newModel;
+      setModel(newModel);
+      clearTimeout(autoSaveTimeoutRef.current);
+      autoSaveTimeoutRef.current = setTimeout(async () => {
+        clearTimeout(autoSaveTimeoutRef.current);
+        save({ modelPath, model: buildJsonSchema(newModel) });
+      }, saveAfterMs);
+    },
+    [modelPath, save]
+  );
```

Comment:
Should we create a debounce util/hook instead? 🤔 Then we can re-use the debounce logic on other input fields a well?